File: src/main/java/toast/client/modules/ModuleManager.java
package toast.client.modules;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.client.MinecraftClient;
import org.lwjgl.glfw.GLFW;
import toast.client.modules.combat.AutoRespawn;
import toast.client.modules.combat.AutoTotem;
import toast.client.modules.combat.BowSpam;
import toast.client.modules.combat.KillAura;
import toast.client.modules.misc.*;
import toast.client.modules.movement.*;
import toast.client.modules.player.AutoTool;
import toast.client.modules.player.Surround;
import toast.client.modules.render.ClickGui;
import toast.client.modules.render.Fullbright;
import toast.client.modules.render.HUD;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

import static toast.client.ToastClient.CONFIG_MANAGER;

@Environment(EnvType.CLIENT)
public class ModuleManager {
    public static final CopyOnWriteArrayList<Module> modules = new CopyOnWriteArrayList<>();

    public void onKey(int key, int action) {
        for (Module module : modules) {
            if (module.getKey() == key && action == GLFW.GLFW_PRESS && MinecraftClient.getInstance().currentScreen == null) {
                if (MinecraftClient.getInstance().inGameHud.getChatHud().isChatFocused()) continue;
                if (!module.getClass().equals(Panic.class) && Panic.IsPanicking()) return;
                module.toggle();
            }
        }
    }

    public Module getModule(Class<? extends Module> moduleClass) {
        for (Module module : modules) {
            if (module.getClass() == moduleClass) {
                return module;
            }
        }
        return null;
    }

    public Module getModule(String name) {
        for (Module module : modules) {
            if (module.getName().equals(name)) {
                return module;
            }
        }
        return null;
    }

    public CopyOnWriteArrayList<Module> getModules() {
        return modules;
    }

    public List<Module> getModulesInCategory(Module.Category category) {
        List<Module> moduleList = new ArrayList<>();
        for (Module module : modules) {
            if (module.getCategory() == category) {
                moduleList.add(module);
            }
        }
        return moduleList;
    }

    public void loadModules() {
        modules.clear();
        // alphabetical order please
        modules.add(new AutoWalk());
        modules.add(new AutoRespawn());
        modules.add(new AutoTool());
        modules.add(new AutoTotem());
        //modules.add(new BlockESP());
        modules.add(new BowSpam());
        modules.add(new CustomChat());
        modules.add(new ClickGui());
        modules.add(new FastStop());
        modules.add(new Flight());
        modules.add(new Fullbright());
        //modules.add(new HoleESP());
        modules.add(new HUD());
        modules.add(new KillAura());
        //modules.add(new NameTags());
        modules.add(new NoFall());
        modules.add(new Panic());
        modules.add(new PortalChat());
        //modules.add(new ShulkerPreview());
        modules.add(new Spammer());
        modules.add(new Sprint());
        //modules.add(new StorageESP());
        modules.add(new Surround());
        modules.add(new Velocity());
        CONFIG_MANAGER.enableWrite();
        CONFIG_MANAGER.loadConfig();
        CONFIG_MANAGER.loadKeyBinds();
        CONFIG_MANAGER.loadModules();
    }
}

File: src/main/java/toast/client/modules/ModuleManager.kt
package toast.client.modules

import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.MinecraftClient
import org.lwjgl.glfw.GLFW
import org.reflections.Reflections
import toast.client.ToastClient
import toast.client.modules.misc.Panic
import toast.client.modules.misc.Panic.Companion.IsPanicking
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList


/**
 * Manages the loading and accessing of module instances
 */
@Environment(EnvType.CLIENT)
class ModuleManager {
    /**
     * Array containing the instances of all the modules
     */
    val modules: CopyOnWriteArrayList<Module> = CopyOnWriteArrayList<Module>()

    /**
     * Checks each module for a key-bind and toggles the module if the key-bind matches the key
     */
    fun onKey(key: Int, action: Int) {
        val iter = modules.iterator()
        while (iter.hasNext()) {
            val next = iter.next()
            if (next.key == key && action == GLFW.GLFW_PRESS && MinecraftClient.getInstance().currentScreen == null) {
                if (MinecraftClient.getInstance().inGameHud.chatHud.isChatFocused) continue
                if (next.javaClass != Panic::class.java && IsPanicking()) return
                next.toggle()
            }
        }
    }

    /**
     * Gets a module from it's class
     */
    fun getModule(moduleClass: Class<out Module?>): Module? = modules.firstOrNull { it.javaClass == moduleClass }

    /**
     * Gets a module from it's name
     */
    fun getModule(name: String): Module? = modules.firstOrNull { it.name == name }

    /**
     * Gets an array of the modules in a category
     */
    fun getModulesInCategory(category: Module.Category): List<Module> {
        val moduleList: MutableList<Module> = ArrayList()
        val iter = modules.iterator()
        while (iter.hasNext()) {
            val next = iter.next()
            if (next.category === category) {
                moduleList.add(next)
            }
        }
        return moduleList
    }

    /**
     * Loads and initializes all of the modules
     */
    fun loadModules() {
        modules.clear()
        val reflections = Reflections("toast.client.modules")
        val moduleClasses =
                reflections.getSubTypesOf(
                        Module::class.java
                )
        val iter = moduleClasses.iterator()
        while (iter.hasNext()) {
            val module = iter.next().getConstructor().newInstance()
            modules.add(module)
        }
        ToastClient.CONFIG_MANAGER.enableWrite()
        ToastClient.CONFIG_MANAGER.loadConfig()
        ToastClient.CONFIG_MANAGER.loadKeyBinds()
        ToastClient.CONFIG_MANAGER.loadModules()
        ToastClient.CONFIG_MANAGER.loadMacros()
    }
}